### PR TITLE
fix(pubsub): Calling events out-of-order

### PR DESCRIPTION
**Pull request by msbrogli**
Fixes https://github.com/HathorNetwork/hathor-core/issues/481.

## Acceptance criteria

1. Fix calling events out-of-order.
2. Fix wallet wrongfully adding utxos of voided transactions.
3. Remove hard-coded hash from `tests/consensus/test_soft_voided.py`.
4. Refactor events: replace `STORAGE_TX_WINNER` and `STORAGE_TX_VOIDED` by `CONSENSUS_TX_UPDATE`.

### CODE REVIEW

File: tests/consensus/test_soft_voided.py [hunk 1]
```diff
@@ -64,7 +66,10 @@ def test_soft_voided(self):
         self.assertEqual({txA.hash}, metaD1.voided_by)
         graphviz.labels[txD1.hash] = 'txD1'
 
-        txD2 = add_custom_tx(manager2, [(txB, 0)])
+        txD2 = gen_custom_tx(manager2, [(txB, 0)])
```
Comment (pedroferreira1) [inline tests/consensus/test_soft_voided.py:69]: Could you use `gen_custom_tx(manager2, [(txB, 0)], inc_timestamp=1)`? Then you don't need the other two lines below.

File: tests/consensus/test_soft_voided2.py [hunk 2]
```diff
@@ -69,44 +81,36 @@ def do_step(self, i, manager1, tx_base):
...
-            bytes.fromhex('aba55041658213b0bfda880045e5a321cb3d0a88cffa2833e620a2ed1ba27451'),
-            bytes.fromhex('43f1782cc7d5702378d067daefea0460cd1976e5479e17accb6f412be8d26ff5'),
-            bytes.fromhex('24fe23d0128c85bc149963ff904aefe1c11b44b0e0cddad344b308b34d1dbac8'),
-            bytes.fromhex('a6b7b21e4020a70d77fd7467184c65bfb2e4443a0b465b6ace05c1b11b82355b'),
-            bytes.fromhex('982e9a2a3f0b1d22eead5ca8cf9a86a5d9edd981ac06f977245a2e664a752e5f'),
-            bytes.fromhex('b6cddd95fe02c035fde820add901545662c7b3ae85010cde5914b86cb0a6505e'),
-            bytes.fromhex('3a84cff789d3263fea4da636da5c54cb0981a4a0caec1579037a4fa06457f8f3'),
-        ])
-        manager1 = self.create_peer(soft_voided_tx_ids=soft_voided_tx_ids)
+    def _run_test(self, simulator, soft_voided_tx_ids, skip_asserts=False):
```
Comment (pedroferreira1) [inline tests/consensus/test_soft_voided2.py:101]: `skip_asserts` is not being used here.

**Review by pedroferreira1 — APPROVED** [review]

File: tests/consensus/test_soft_voided2.py (hunk 2 above)
Comment (msbrogli) [reply tests/consensus/test_soft_voided2.py:101]: Nice catch! Thanks!

File: tests/consensus/test_soft_voided.py (hunk 1 above)
Comment (msbrogli) [reply tests/consensus/test_soft_voided.py:69]: Nice catch! Fixed! Thanks!

File: tests/consensus/test_soft_voided2.py (hunk 2 above)
Comment (msbrogli) [reply tests/consensus/test_soft_voided2.py]: Fixed!

File: hathor/transaction/storage/transaction_storage.py [hunk 3]
```diff
@@ -407,6 +401,7 @@ def get_best_block_tips(self, timestamp: Optional[float] = None, *, skip_cache:
             elif meta.score > best_score:
                 best_score = meta.score
                 best_tip_blocks = [block_hash]
+        self._best_block_tips = best_tip_blocks[:]
```
Comment (pedroferreira1) [inline hathor/transaction/storage/transaction_storage.py:404]: Should it be `self. _best_block_tips_cache`?

File: hathor/consensus.py [hunk 4]
```diff
@@ -1006,6 +1009,7 @@ def add_voided_by(self, tx: Transaction, voided_hash: bytes) -> bool:
             if tx2.is_block:
                 assert isinstance(tx2, Block)
                 self.context.block_algorithm.mark_as_voided(tx2)
+                tx2.storage.update_best_block_tips_cache(None)
```
Comment (pedroferreira1) [inline hathor/consensus.py:1012]: Why do you need this change? Was it a bug?

File: hathor/transaction/storage/transaction_storage.py (hunk 3 above)
Comment (msbrogli) [reply hathor/transaction/storage/transaction_storage.py:404]: Sure! Great catch! Thanks!

File: hathor/consensus.py (hunk 4 above)
Comment (msbrogli) [reply hathor/consensus.py:1012]: Yes, the best block tips cache wasn't being updated and it was returning the wrong best block in a specific situation. I created this issue to improve our testing tools: https://github.com/HathorNetwork/hathor-core/issues/491.

**Review by pedroferreira1 — APPROVED** [review]

**Review by jansegre — APPROVED** [review]

**Pull request merged**